test/e2e/azure.go

```diff
@@ -0,0 +1,223 @@
+	return nil
+}
+
+// Credentials should in as json string for azure provider.
+func (a *azureTestHelper) makeCredentialsSecret(namespace string) *corev1.Secret {
+	credData := struct {
+		TenantID       string `json:"tenantId"`
+		SubscriptionID string `json:"subscriptionId"`
+		ResourceGroup  string `json:"resourceGroup"`
+		ClientID       string `json:"aadClientId"`
+		ClientSecret   string `json:"aadClientSecret"`
+	}{
+		TenantID:       a.config.TenantID,
+		SubscriptionID: a.config.SubscriptionID,
+		ResourceGroup:  a.config.ResourceGroup,
+		ClientID:       a.config.ClientID,
+		ClientSecret:   a.config.ClientSecret,
+	}
+	azureCreds, err := json.Marshal(credData)
+	if err != nil {
+
+	}
+
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("azure-config-file-%s", randomString(16)),
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			"azure.json": azureCreds,
+		},
+	}
+}
+
+func (a *azureTestHelper) platform() string {
+	return string(configv1.AzurePlatformType)
+}
+
+func (a *azureTestHelper) ensureHostedZone(rootDomain string) (string, []string, error) {
+	location := "global"
+	z, err := a.zoneClient.CreateOrUpdate(context.TODO(), a.config.ResourceGroup, rootDomain,
+		dns.Zone{Location: &location}, "", "")
+	if err != nil {
+		return "", []string{}, err
+	}
+	var zoneID string
+	zoneID = *z.ID
+	nameservers := append(*z.ZoneProperties.NameServers)
+	a.zoneName = rootDomain
+	return zoneID, nameservers, nil
+}
+
+func (a *azureTestHelper) deleteHostedZone(rootDomain string) error {
+	if a.zoneName == "" {
+		fmt.Printf("ZoneName is empty, nothing to be deleted")
+		return nil
+	}
+	if _, err := a.zoneClient.Delete(context.TODO(), a.config.ResourceGroup, a.zoneName, ""); err != nil {
+		return fmt.Errorf("unable to delete zone :%s, failed error: %v", a.zoneName, err)
+	}
+	// verify the zone is present
+	if _, err := a.isZoneNameExists(); err != nil {
+		return fmt.Errorf("unable to verfy zone deletion,failed with error : %v", err)
+	}
+
+	return nil
+}
+
+func (a *azureTestHelper) isZoneNameExists() (bool, error) {
+	ctx := context.TODO()
+	zonesIterator, err := a.zoneClient.ListByResourceGroupComplete(ctx, a.config.ResourceGroup, nil)
+	if err != nil {
+		return false, err
+	}
+
+	for zonesIterator.NotDone() {
+		zone := zonesIterator.Value()
+
+		if zone.Name != nil && a.zoneName == *zone.Name {
+			return true, nil
+		}
+		err = zonesIterator.NextWithContext(ctx)
+		if err != nil {
+			return false, err
+		}
+	}
+	return false, nil
+}
+
+// ref: https://github.com/kubernetes-sigs/external-dns/blob/master/provider/azure/azure.go
+// getAccessToken retrieves Azure API access token.
+func getAccessToken(cfg *cluserConfig) (*adal.ServicePrincipalToken, error) {
+	// Try to retrieve token with service principal credentials.
+	// Try to use service principal first, some AKS clusters are in an intermediate state that `UseManagedIdentityExtension` is `true`
```

Comment:
Not sure this comment is still valid as `UseManagedIdentityExtension` seems to be a leftover.

Reply:
ohh, didnt see the comment section, will take out now.